src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```diff
@@ -0,0 +1,177 @@
+using Microsoft.Azure.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal sealed class CallerClientResultsManager : ICallerClientResultsManager, IInvocationBinder
+    {
+        private readonly ConcurrentDictionary<string, PendingInvocation> _pendingInvocations = new();
+        private readonly ConcurrentDictionary<string, List<string>> _serviceMappingMessages = new();
+        private readonly string _serverGUID = Guid.NewGuid().ToString();
+        private long _lastInvocationId = 0;
+
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public string GenerateInvocationId(string connectionId)
+        {
+            return $"{connectionId}-{_serverGUID}-{Interlocked.Increment(ref _lastInvocationId)}";
+        }
+
+        public Task<T> AddInvocation<T>(string connectionId, string invocationId, string instanceId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(
+                cancellationToken,
+                () => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
+
+            var result = _pendingInvocations.TryAdd(invocationId, new PendingInvocation(typeof(T), connectionId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            // When the caller server is also the client router, Azure SignalR service won't send the ServiceMappingMessage to server.
+            // To handle this condition, CallerClientResultsManager should record this mapping information itself without receiving a ServiceMappingMessage from service.
+            if (instanceId != null)
+            {
+                AddServiceMappingMessage(new ServiceMappingMessage(connectionId, invocationId, instanceId));
+            }
+
+            return tcs.Task;
+        }
+
+        public void AddServiceMappingMessage(ServiceMappingMessage serviceMappingMessage)
+        {
+            _serviceMappingMessages.AddOrUpdate(
+                serviceMappingMessage.InstanceId,
+                new List<string>() { serviceMappingMessage.InvocationId },
+                (key, valueList) => { valueList.Add(serviceMappingMessage.InvocationId); return valueList; });
```

Comment:
this `valueList.Add` is not thread safe

Reply:
I replace `List<string>` with `ConcurrentBag<string>`
Now `private readonly ConcurrentDictionary<string, ConcurrentBag<string>> _serviceMappingMessages = new();`